### PR TITLE
LUCENE-10493: factor out Viterbi algorithm and share it between kuromoji and nori

### PR DESCRIPTION
**Pull request by mocobeta**
See also https://issues.apache.org/jira/browse/LUCENE-10493.

Main changes:
- Add a utility class `org.apache.lucene.analysis.morph.Viterbi`, which performs the Viterbi algorithm for Hidden Markov Model (or Conditional Random Fields) based tokenizers.
  - The most important interfaces of this class are `forward()` (read one character from the input and grow the graph) and `backward()` (add output tokens on the minimum cost path to the pending list that is consumed by `Tokenizer#incrementToken()`).
- Add extensions of `Viterbi` to Kuromoji and Nori.
   - Note: Those extensions are needed because the current `backtrace()` in `JapaneseTokenizer` and `KoreanTokenizer` needs language-specific information. We could (and should) decouple each language-related stuff from `backtrace()`, but it'd be not a trivial refactoring to me. I'd defer it for now.
   - Note: The extension for Kuromoji also performs n-best cost path calculation. It would be great if we decouple this from JapaneseTokenizer but I found the current `backtraceNBest()` needs information that is specific to JapaneseToeknizer... I'd defer the generalization for now.
- Add a parameterized `org.apache.lucene.analysis.morph.GraphvizFormatter` class to analysis-common, which is shared between Kuromoji and Nori.
   - `org.apache.lucene.analysis.ja.GraphvizFormatter` and `org.apache.lucene.analysis.ko.GraphvizFormatter` are removed.
- Since all the segmentation work is now done by `Viterbi`, `JapaneseTokenizer` and `KoreanTokenizer` are facade classes that only have constructors and `incrementToken()`, `close()`, `reset()`, `end()`.

### CODE REVIEW

**Comment by mocobeta** [comment]
@rmuir would you mind reviewing this, or do you think we shouldn't proceed this way?
I know it's a bit radical refactoring but I cannot come up with a better way to unify the two tokenizers than this, sorry... let me know if it'd be better to stop pursuing this way.

**Comment by rmuir** [comment]
Sure, sorry for the slow response.

**Comment by rmuir** [comment]
I think @mikemccand actually created most of this code and is most familiar with it. Mike, if you have time can you look too?

For the special n-best class, is the issue that `nori` simply doesn't offer this n-best feature? For the future, I wonder if there is some reason it doesn't make sense there too? Maybe it was just overlooked before?

**Comment by mikemccand** [comment]
Whoa, this sounds awesome!  I will try to review soon.  Thanks @mocobeta.

**Comment by mocobeta** [comment]
Hi Robert and Mike, thank you for your response. I think this can be kept open for a sufficient time period - it is unlikely to happen large conflicts between these changes and the main branch.

I reviewed it several times during making this patch and I believe this does not change the tokenizers' behavior (although I can't fully guarantee that). I'd be glad if you give feedback about the overall design, and ideas to make the interfaces safer for further refactoring or development.

File: lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/ViterbiNBest.java [hunk 1]
```diff
@@ -0,0 +1,1253 @@
...
+  @Override
+  protected void backtrace(Position endPosData, int fromIDX) throws IOException {
+    final int endPos = endPosData.getPos();
+
+    /**
+     * LUCENE-10059: If the endPos is the same as lastBackTracePos, we don't want to backtrace to
+     * avoid an assertion error {@link RollingCharBuffer#get(int)} when it tries to generate an
+     * empty buffer
+     */
+    if (endPos == lastBackTracePos) {
```
Comment (mikemccand) [inline lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/ViterbiNBest.java:204]: I wonder if this was a bug fix to Kuromoji that should've been ported over to Nori?

**Review by mikemccand — APPROVED** [review]
Thank you @mocobeta for tackling this crazy complex code!!

Big +1 to take progress-not-perfection approach and factor out code that is clearly in common into `Viterbi` base-class.

We can later try to go further e.g. bringing n-best to Nori as well.  I peeked at the methods like `backtrace` and they really are quite divergent across the two, and I added a comment on one possible bug-fix that we failed to cross-port.  So it's awesome to take this step today to prevent further unnecessary divergence, and to enable future tokenizers to more easily re-use the Viterbi algorithm for finding best 1 or N tokenization paths for weighted dictionary based tokenizers.

I love how much common code you were able to delete!  And how `GraphvizFormatter` is also factored out :)

File: lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/ViterbiNBest.java (hunk 1 above)
Comment (mocobeta) [reply lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/ViterbiNBest.java:204]: +1 I think this should be ported to Nori.

**Review by mikemccand — APPROVED** [review]
Looks awesome, thanks @mocobeta!

**Comment by mocobeta** [comment]
@mikemccand thanks for reviewing such a huge PR!

I'll look through this again, and merge it maybe next week.
@rmuir if you have any comments I'll try to reflect the feedback on this, thanks.

**Comment by mocobeta** [comment]
I'm merging this and will observe the nightly benchmark (this shouldn't affect the analysis performance).

**Comment by mocobeta** [comment]
I looked closely at the code again and then found the n-best logic can be separated from JapaneseTokenizer (not perfect though); opened a follow-up #846. It's a safe change I think and I plan to merge this to main if there are no objections.

**Pull request merged**